pkg/services/ngalert/eval/eval.go

```diff
@@ -302,10 +302,11 @@ type NumberValueCapture struct {
 }
 
 func queryDataResponseToExecutionResults(c models.Condition, execResp *backend.QueryDataResponse) ExecutionResults {
-	// eval captures for the '__value_string__' annotation and the Value property of the API response.
-	captures := make([]NumberValueCapture, 0, len(execResp.Responses))
-	captureVal := func(refID string, labels data.Labels, value *float64) {
-		captures = append(captures, NumberValueCapture{
+	// captures contains the values of all instant queries and expressions for each dimension
+	captures := make(map[data.Fingerprint][]NumberValueCapture, len(execResp.Responses))
```

Comment:
Took me a while to understand the effect of pre-allocating the map here, and if it actually helps:

- Initially it looked wrong sizing the map based on the number of responses, as grouping would imply there will be fewer entries in the map than there are responses. If the labels of the responses are mostly unique, then there will nearly always be the same number of entries in the map as responses.
- This means however, that mostly we will be allocating slices to contain a single entry, and so any allocations saved by sizing the map will be dwarfed by allocating N slices instead.

I don't think this worth being concerned about right now, as the algorithmic improvement is worth the extra allocations, but there might be a time where we optimize this further.